lib/chat_bot.ex

```diff
@@ -126,7 +126,25 @@ defmodule Telegram.ChatBot do
               | {:ok, next_chat_state :: chat_state(), timeout :: timeout()}
               | {:stop, next_chat_state :: chat_state()}
 
-  @optional_callbacks handle_resume: 1, handle_info: 4, handle_timeout: 3
+  @doc """
+  Allows a chat bot to customize how updates are handled by this bot.
+
+  A return value of {:ok, %{"id" => id, ...}} will result in a new instance
+  of the bot being spun up to handle the update as a full chat.
+
+  A return value of {:transient, %{"id" => id, ...}} will result in a transient
+  instance of the bot being spun up to handle the update. In this case the process
+  will only exist to handle this specific update.
+
+  A return value of nil causes the update to be ignored.
+
+  This callback is optional and if not implemented the legacy behavior will
+  be used. The returned chat must have at least an entry containing the chat
+  id
+  """
+  @callback get_chat(update_type :: String.t(), update :: Types.update()) :: {:ok, Types.chat()} | {:transient, Types.chat()} | nil
```

Comment:
As we said, let's make this agnostic about "transient".

Maybe:

```elixir
@callback get_chat(update :: Types.update()) :: {:ok, ChatBot.id()} | :ignore
```

Note:
- I'm passing the whole update the the callback. The user implementation can simply patter match on the updated`%{""message" => _}` to identify the updates type they want to manage.
- I'm returning `{:ok, _} | :ignore` to be more idiomatic following the line of `Supervisor.on_start()`
- I'm returning `{:ok, ChatBot.id()}` where I would like `ChatBot.id()` to be a `String.t()` the callback can construct to identify a unique bot instance. Innmost cases this can simply be the `update.message.chat.id`, `inline_query.id`, etc. The reason for this is to let the users to be in full control of the instances and hence how they identify (and derive the identifier) of the chat (there maybe different business logic).
 
 Feel free to change accordingly the `Telegram.Bot.Utils.get_chat()` signature and break the backward-compatibility. I think I'll bump the major version with this PR anyway.

Reply:
Make perfect sense!

I like option 2)

```elixir

  defmodule Chat do
    @moduledoc """
    ...
    """

    @enforce_keys [:id]
    defstruct @enforce_keys ++ [metadata: %{}]
    
    @type %__MODULE__{
      id: String.t(),
      metadata: map()
    }
  end
  
  @doc """
  Invoked once when the chat starts.
  Return the initial chat_state.
  """
  @callback init(chat :: Chat.t()) ::
              {:ok, initial_state :: chat_state()}
              | {:ok, initial_state :: chat_state(), timeout :: timeout()}
```

It looks explicit and enforce a clear link between `get_chat()` and `init()`.

Go for it!